tests/cnf/core/network/sriov/tests/qinq.go

```diff
@@ -663,3 +796,132 @@ func disableQinQOnSwitch(switchCredentials *sriovenv.SwitchCredentials, switchIn
+
+	return dpdkPod
+}
+
+func defineAndCreateClientDPDKPod(
+	podName,
+	nodeName string,
+	serverPodNetConfig []*multus.NetworkSelectionElement) *pod.Builder {
+	var rootUser int64
+	securityContext := corev1.SecurityContext{
+		RunAsUser: &rootUser,
+		Capabilities: &corev1.Capabilities{
+			Add: []corev1.Capability{"IPC_LOCK", "SYS_RESOURCE", "NET_RAW"},
+		},
+	}
+
+	dpdkContainerCfg, err := pod.NewContainerBuilder(podName, NetConfig.DpdkTestContainer,
+		[]string{"/bin/bash", "-c", "sleep INF"}).WithSecurityContext(&securityContext).
+		WithResourceLimit("2Gi", "1Gi", 4).
+		WithResourceRequest("2Gi", "1Gi", 4).WithEnvVar("RUN_TYPE", "testcmd").GetContainerCfg()
+	Expect(err).ToNot(HaveOccurred(), "Fail to define client dpdk container")
+
+	dpdkPod := pod.NewBuilder(APIClient, podName, tsparams.TestNamespaceName,
+		NetConfig.DpdkTestContainer).WithSecondaryNetwork(serverPodNetConfig).
+		DefineOnNode(nodeName).RedefineDefaultContainer(*dpdkContainerCfg).WithHugePages()
+
+	dpdkPod, err = dpdkPod.CreateAndWaitUntilRunning(4 * time.Minute)
+	Expect(err).ToNot(HaveOccurred(), "Fail to create a dpdk client pod")
+
+	return dpdkPod
+}
+
+func rxTrafficOnClientPod(clientPod *pod.Builder) {
+	clientRxCmd := defineTestClientPmdCmd(tsparams.ClientMacAddress, "${PCIDEVICE_OPENSHIFT_IO_SRIOVPOLICYVFIOPCI}")
+
+	By("Validate dpdk_testpmd traffic from the server to the client using CVLAN100.")
+
+	clientOut, err := clientPod.ExecCommand(clientRxCmd)
+
+	var timeoutError = "command terminated with exit code 137"
+	// The timeout -s SIGKILL 20 dpdk-testpmd cause the following error after a success command execution.
+	if err.Error() != timeoutError {
+		Expect(err).ToNot(HaveOccurred(), "Fail to exec cmd")
+	}
+
+	By("Parsing output from the DPDK application")
+	glog.V(90).Infof("Processing testpdm output from client pod")
+	Expect(checkRxOnly(clientOut.String())).Should(BeTrue(),
+		fmt.Sprintf("Fail to process output from dpdk application %s", clientOut.String()))
+}
+
+// checkRxOnly checks the number of incoming packets.
+func checkRxOnly(out string) bool {
+	lines := strings.Split(out, "\n")
+	Expect(len(lines)).To(BeNumerically(">=", 3),
+		"Fail line list contains less than 3 elements")
+
+	for i, line := range lines {
+		if strings.Contains(line, "NIC statistics for port") {
+			if len(lines) > i && getNumberOfPackets(lines[i+1], "RX") > 0 {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// getNumberOfPackets counts the number of packets in the lines with RX output.
+func getNumberOfPackets(line, firstFieldSubstr string) int {
+	splitLine := strings.Fields(line)
+	Expect(splitLine[0]).To(ContainSubstring(firstFieldSubstr), "Fail to find expected substring")
+	Expect(len(splitLine)).To(Equal(6), "the slice doesn't contain 6 elements")
+	numberOfPackets, err := strconv.Atoi(splitLine[1])
+	Expect(err).ToNot(HaveOccurred(), "Fail to convert string to integer")
+
+	return numberOfPackets
+}
```

Comment:
I see many functions similar to the ones in rootless dpdk. Is it possible to have common functions for both of them?

Reply:
I moved the common code higher in the test suite.